modules/ROOT/pages/traversal-framework/traversal_framework_java_api.adoc

```diff
@@ -0,0 +1,353 @@
+     td = tx.traversalDescription();
+            .uniqueness( Uniqueness.RELATIONSHIP_GLOBAL )
+}
+
+td.traverse( startNode );
+----
+
+For an example demonstrating how to use uniqueness in a traversal see <<examples-uniqueness-of-paths-in-traversals, Uniqueness of Paths in traversals>>.
+
+[[traversal-java-api-branchselector]]
+== BranchSelector
+
+A `BranchSelector` / `BranchOrderingPolicy` is used for selecting which branch of the traversal to attempt next.
+This is used for implementing traversal orderings.
+
+The Traversal Framework provides a few basic ordering implementations based on the link:https://en.wikipedia.org/wiki/Depth-first_search[depth-first^] and link:https://en.wikipedia.org/wiki/Breadth-first_search[breadth-first^] algorithms:
+
+* `BranchOrderingPolicies.PREORDER_DEPTH_FIRST` -- Traversing depth first, visiting each node before visiting its child nodes.
+* `BranchOrderingPolicies.POSTORDER_DEPTH_FIRST` -- Traversing depth first, visiting each node after visiting its child nodes.
+* `BranchOrderingPolicies.PREORDER_BREADTH_FIRST` -- Traversing breadth first, visiting each node before visiting its child nodes.
+* `BranchOrderingPolicies.POSTORDER_BREADTH_FIRST` -- Traversing breadth first, visiting each node after visiting its child nodes.
+
+[NOTE]
+====
+Breadth-first traversals have a higher memory overhead than depth-first traversals.
+====
+
+A `BranchSelector` carries state and hence needs to be uniquely instantiated for each traversal.
+Therefore it is supplied to the `TraversalDescription` through a `BranchOrderingPolicy` interface, which is a factory of `BranchSelector` instances.
+
+A user of the Traversal Framework rarely needs to implement their own `BranchSelector` or `BranchOrderingPolicy`, it is provided to let graph algorithm implementors provide their own traversal orders.
+The Neo4j Graph Algorithms package contains for example a `BestFirst` order `BranchSelector` / `BranchOrderingPolicy` that is used in BestFirst search algorithms such as A* and Dijkstra.
+
+=== Example
+Given the following graph, these are the results of the branch ordering policies without any extra filter:
+
+image::traversal_order_example_graph.png[align="center", role="middle", width=200]
+
+[cols="1,1"]
+|===
+|Ordering policy |Order of the nodes in traversal
+
+|`BranchOrderingPolicies.PREORDER_DEPTH_FIRST`
+|a, b, d, c, e
+
+|`BranchOrderingPolicies.POSTORDER_DEPTH_FIRST`
+|d, b, e, c, a
+
+|`BranchOrderingPolicies.PREORDER_BREADTH_FIRST`
+|a, b, c, d, e
+
+|`BranchOrderingPolicies.POSTORDER_BREADTH_FIRST`
+|d, e, b, c, a
+|===
+
+=== BranchOrderingPolicy
+
+A `BranchOrderingPolicy` is a factory for creating ``BranchSelector``s to decide in what order branches are returned
+(where a branch's position is represented as a `Path` from the start node to the current node).
+
+Depth-first and breadth-first are common policies and can be accessed by the convenience methods breadthFirst() / depthFirst().
```

Comment:
Would it make sense to lead on the convenience methods as they are the most likely to be used and the simplest to understand?

Reply:
We think it is nice to introduce what breadth first/ depth first is before jumping in to methods